### PR TITLE
WIP "Struct" branch

### PR DESCRIPTION
**Pull request by cjordan**
Hey @derwentx, I've left all my thoughts in the WIP commit's message. Let me know what you think, if you have the time. FTR, this should not be merged yet!

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -3,16 +3,16 @@
 #![warn(clippy::missing_safety_doc)]
 #![warn(clippy::missing_errors_doc)]
 
-//! Birli is a library of common preprocessing tasks performed in the data pipeline of the Murchison
-//! Widefield Array (MWA) Telescope.
+//! Birli is a library of common preprocessing tasks performed in the data
+//! pipeline of the Murchison Widefield Array (MWA) Telescope.
```
Comment (derwentx) [inline src/lib.rs:7]: what formatter plugin are you using? It'd be good to sync up formatter settings

File: src/lib.rs [hunk 2]
```diff
@@ -97,13 +98,13 @@
...
-//! For more details about AOFlagger's interface, check out the [`aoflagger::AOFlagger`]
-//! documentation
+//! Birli reads visibilities with [`MWALib`] and uses CXX to bind to the
+//! [`AOFlagger`] C++ library. For more details about AOFlagger's interface,
+//! check out the [`aoflagger::AOFlagger`] documentation
 //!
-//! [`MWALib`]: https://github.com/MWATelescope/mwalib
-//! [`AOFlagger`]: https://gitlab.com/aroffringa/aoflagger
-//! [`aoflagger::AOFlagger`]: http://www.andreoffringa.org/aoflagger/doxygen/classaoflagger_1_1AOFlagger.html
+//! [`MWALib`]: https://github.com/MWATelescope/mwalib [`AOFlagger`]:
```
Comment (derwentx) [inline src/lib.rs:105]: yup, looks like the formatter can't handle rustdoc links https://doc.rust-lang.org/rustdoc/linking-to-items-by-name.html

**Comment by derwentx** [comment]
nice! I would love to include these in the context too (probably a later commit)
```
        timestep_range: &Range<usize>,
        coarse_chan_range: &Range<usize>,
        baseline_idxs: &[usize],
```

**Comment by derwentx** [comment]
BTW I've seen that error in "generate coverage report" before, this is how i fixed it. https://github.com/MWATelescope/Birli/commit/564877cbfd1b88b9fe445a47c61ccc2a1a430df3

**Comment by derwentx** [comment]
> nice! I would love to include these in the context too (probably a later commit)
> 
> ```
>         timestep_range: &Range<usize>,
>         coarse_chan_range: &Range<usize>,
>         baseline_idxs: &[usize],
> ```

now that I think about this more, our eventual aim is to consolidate everything related to file format i/o into Marlu / mwa_rust_core, so these don't belong in this particular layer of our matryoshka of contexts

**Comment by cjordan** [comment]
We probably should have an in-person conversation, but another thought I had on my approach here was to have `from_mwalib` and `from_files` methods instead of `new` -- this way, a user could make the `BirliContext` from an existing mwalib context or from the files that would've made it.

I haven't configured my rustfmt at all, so I'll be interested to see how yours is set up!

**Comment by derwentx** [comment]
> We probably should have an in-person conversation, but another thought I had on my approach here was to have `from_mwalib` and `from_files` methods instead of `new` -- this way, a user could make the `BirliContext` from an existing mwalib context or from the files that would've made it.

Sounds very reasonable!

> I haven't configured my rustfmt at all, so I'll be interested to see how yours is set up!

Easiest thing to do would be to create an [editorconfig](https://editorconfig.org/) that we both agree on. It looks like the only difference is your formatter's column width and handling of rustdoc comments.

**Comment by d3v-null** [comment]
I think I've covered everything from this branch in #55 , let me know if there's anything missing @cjordan

**Comment by d3v-null** [comment]
closed by https://github.com/MWATelescope/Birli/pull/55

**Pull request closed without merging**